internal/util/util.go

```diff
@@ -219,7 +219,61 @@ func FSize(f *os.File) int64 {
 // IsWordChar returns whether or not the string is a 'word character'
 // Word characters are defined as numbers, letters, or '_'
 func IsWordChar(r rune) bool {
-	return unicode.IsLetter(r) || unicode.IsNumber(r) || r == '_'
+	return IsAlphanumeric(r) || IsWordDelimiter(r)
+}
+
+// IsUpperWordChar returns whether or not the string is a 'upper word character'
+// Upper word characters are defined as numbers, upper-case letters, or delimiters ([A-Z0-9_])
+func IsUpperWordChar(r rune) bool {
+	return IsUpperAlphanumeric(r) || IsWordDelimiter(r)
+}
+
+// IsLowerWordChar returns whether or not the string is a 'lower word character'
+// Lower word characters are defined as numbers, upper-case letters, or delimiters ([a-z0-9_])
+func IsLowerWordChar(r rune) bool {
+	return IsLowerAlphanumeric(r) || IsWordDelimiter(r)
+}
+
+// IsWordDelimiter returns whether or not the string is a 'delimiter character'
+// Delimiter characters are defined as '_'
+func IsWordDelimiter(r rune) bool {
+	return r == '_'
+}
+
+// IsNoneWordChar returns whether or not the string is not a 'word character'
+// None word characters are defined as all characters not being alphanumerics, whitespaces or delimiters ([!A-Za-z0-9\s_])
+func IsNoneWordChar(r rune) bool {
```

Comment:
This also seems rather unneeded. What's wrong with explicit `!IsWordChar() && !IsWhitespace()` checks?

Moreover, there is already `IsNonAlphaNumeric()`, i.e. a function that uses almost the same naming style but different semantics: it returns true for whitespace. So it you add these functions but leave the existing `IsNonAlphaNumeric()` as is, IMO that creates a mess.

And there is a reason why `IsNonAlphaNumeric()` is a separate function: it is used as a callback, in multiple places. Whereas in your case there is no strong need for a separate function.

So I'd propose to:
1. Use explicit `!IsWordChar() && !IsWhitespace()` checks instead of adding `IsNoneWordChar()`
2. Rename the existing `IsNonAlphaNumeric()` to `IsNonWordChar()`.

Reply:
So, in `buffer/cursor.go` I would need to change 2 times 2 references of `IsUpperLetter`.

One change would look like to following, from:

```go
if util.IsUpperLetter(c.RuneUnder(c.X)) &&
	util.IsUpperLetter(c.RuneUnder(c.X+1)) {
	for util.IsUpperAlphanumeric(c.RuneUnder(c.X)) {
		if c.X == util.CharacterCount(c.buf.LineBytes(c.Y)) {
			return
		}
		c.Right()
	}
	if util.IsLowerAlphanumeric(c.RuneUnder(c.X)) {
		c.Left()
	}
}
```

to
```go
import(
+ "unicode"  # an additional import is now required
)

...

r := c.RuneUnder(c.X)
next_r := c.RuneUnder(c.X + 1)
if (unicode.IsLetter(r) && unicode.IsUpper(r)) &&
	(unicode.IsLetter(next_r) && unicode.IsUpper(next_r)) {
	for util.IsUpperAlphanumeric(c.RuneUnder(c.X)) {
		if c.X == util.CharacterCount(c.buf.LineBytes(c.Y)) {
			return
		}
		c.Right()
	}
	if util.IsLowerAlphanumeric(c.RuneUnder(c.X)) {
		c.Left()
	}
} 
```

Is that really desired? We now have a bunch of methods to classify runes, why not elaborate on this concept and just use those?

Reply:
Regarding the `IsNonAlphaNumeric` function:
Absolutely right, I guess this was forgotten or subject to discussion, not sure anymore.

I would suggest the following:
	1) Rename all functions containing `None` to contain `Non` instead. e.g. `IsNoneWordChar` -> `IsNonWordChar`
	2) Change `IsNonAlphaNumeric` from
```go
	func IsNonAlphaNumeric(c rune) bool {
		return !unicode.IsLetter(c) && !unicode.IsNumber(c) && c != '_'
	}
```
to
```go
	func IsNonAlphaNumeric(c rune) bool {
		return !unicode.IsLetter(c) && !unicode.IsNumber(c) && !IsWordDelimiter(c)
	}
```

Reply:
> So, in `buffer/cursor.go` I would need to change 2 times 2 references of `IsUpperLetter`. [...]

This is probably a reply to another comment?

> I would suggest the following:
> 1) Rename all functions containing None to contain Non instead. e.g. IsNoneWordChar -> IsNonWordChar
> 2) Change IsNonAlphaNumeric from [...]

That would still leave it inconsistent. `IsNonAlphaNumeric()` would still unexpectedly return false for `_` (whereas you also add `IsAlphaNumeric()` which also, expectedly, returns false for `_`). So why not just rename `IsNonAlphaNumeric()` to `IsNonWordChar()`?

Again, what exactly is the point of your `IsNoneWordChar()` instead of explicit `!IsWordChar() && !IsWhitespace()`? We don't need to add wrappers for everything in the world.

Reply:
> Again, what exactly is the point of your IsNoneWordChar() instead of explicit !IsWordChar() && !IsWhitespace()? We don't need to add wrappers for everything in the world.

Simplicity and following the design of the original implementation.

When I implemented this feature I tried to stick to original implementation as close as possible.
I realized that for rune classification there were some functions available in `util.go`, acting as classification wrappers.  Another indicator for this is that there is no `import "unicode" ` statement in `cursor.go`. A file consisting of ~500 lines and you cannot find a single `unicode.` call? I assumed using the wrappers from `util.go`  is intentional and therefore the way to go. So, extending them felt natural to me.

Have a look on `IsWhitespace()` in `util.go`:
```go

// IsWhitespace returns true if the given rune is a space, tab, or newline
func IsWhitespace(c rune) bool {
	return unicode.IsSpace(c)
}
```

It is just wrapping the `unicode.` call and I just followed this implicit design.
So, given this preexisting concept, is it okay to introduce `IsNoneWordChar()` just for wrapping 2 other checks?
Considering the `IsWhitespace()` function I would say yes.
It gets around importing `unicode` and simplifies the code, by checking the two things need for this particular rune classification instead of having 2 checks here are there all over the code.

Again, I know exactly why you are proposing this. In quite some cases programmers tend to over-engineer, by adding more and more wrappers. But imho this is not the case here.

I am still thinking about the `IsNonAlphaNumeric` vs.  `IsAlphanumeric` vs. `IsNoneWordChar` thing...

You are right about `IsNonAlphaNumeric()` not being a counterpart to `IsAlphanumeric()`.
So, I guess we could rename `IsNonAlphaNumeric()` to `IsNonWordChar()`, but I need the whitespace check which is not in the original version. Which leaves us with 2 options:

1) Create a new utils function
```go
// this should be equivalent to the original version
func IsNoneWordChar(c rune) bool {
	return !IsAlphanumeric(c) && !IsWordDelimiter(c)
}

// this is the new helper
func IsNoneWordCharNorWhitespace(c rune) bool {
	return IsNoneWordChar(c) && !IsWhitespace(c)
}

// So, that we can use:
...
	if util.IsNoneWordCharNorWhitespace(c.RuneUnder(c.X)) {
		for util.IsNoneWordCharNorWhitespace(c.RuneUnder(c.X)) {
			if c.X == util.CharacterCount(c.buf.LineBytes(c.Y)) {
				c.Right()
				return
			}
			c.Right()
		}
		return
	}
```

2) Do not create a new helper function, just use `IsNoneWordChar()`  and `!IsWhitespace()`

```go
	if util.IsNoneWordChar(c.RuneUnder(c.X)) && !util.IsWhitespace(c.RuneUnder(c.X)) {
		for util.IsNoneWordChar(c.RuneUnder(c.X)) && !util.IsWhitespace(c.RuneUnder(c.X)) {
			if c.X == util.CharacterCount(c.buf.LineBytes(c.Y)) {
				c.Right()
				return
			}
			c.Right()
		}
		return
	}

```

I really hope I got all your points ... and didn't forget anything.

Reply:
I don't think it would be a problem if we added `import "unicode"` to `cursor.go`. Anyway, for `!IsWordChar() && !IsWhitespace()` you don't need `import "unicode"`, you are using wrappers already.

> ```go
> // this should be equivalent to the original version
> func IsNoneWordChar(c rune) bool {
> 	return !IsAlphanumeric(c) && !IsWordDelimiter(c)
> }
>```

Or just:

```go
func IsNonWordChar(c rune) bool {
	return !IsWordChar(c)
}
```

BTW, s/IsNone/IsNon/ ?

>     2. Do not create a new helper function, just use `IsNoneWordChar()`  and `!IsWhitespace()`
> [...]

I'd vote for that. IMHO `IsNoneWordCharNorWhitespace()` sounds just crazy (which is a good thing: it doesn't conceal that it is crazy).

Reply:
I had a slight feeling you would go with option 2 :smile: . But that's alright, lets do it this way.

> BTW, s/IsNone/IsNon/ ?

I really don't mind. What makes more sense to you?


> IsNoneWordCharNorWhitespace() sounds just crazy ...

Yep, :smile:, I have to admit i didn't spend much time on that name ...

Reply:
> > BTW, s/IsNone/IsNon/ ?
> 
> I really don't mind. What makes more sense to you?

IsNon. The word `none` means something different.